File: Assignment4.js

class EmployeeScheduler {
    
    constructor() {
        this.employees = {};
        this.schedule = {};
        this.days = ["Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday", "Sunday"];
        this.shifts = ["morning", "afternoon", "evening"];
        this.maxEmployees = 6;
        this.initializeSchedule();
    }

    

    addEmployee(name, preferences) {
        if (Object.keys(this.employees).length < this.maxEmployees && !this.employees[name]) {
            this.employees[name] = { preferences, assignedDays: 0 };
            this.assignShifts();
            this.resolveConflicts();
            this.displaySchedule();
        } else {
            alert("No employee works more than one shift per day!(Also every employee works only 5 days per week)");
            
        }
    }

    initializeSchedule() {
        this.days.forEach(day => {
            this.schedule[day] = {
                morning: [],
                afternoon: [],
                evening: []
            };
        });
    }

    
    assignShifts() {
        for (const day of this.days) {
            for (const [name, data] of Object.entries(this.employees)) {
                if (data.assignedDays < 5) {
                    let assigned = false;
                    for (const shift of data.preferences) {
                        if (this.schedule[day][shift].length < 2) {
                            this.schedule[day][shift].push(name);
                            data.assignedDays++;
                            assigned = true;
                            break;
                        }
                    }
                    
                    if (!assigned) {
                        for (const shift of this.shifts) {
                            if (this.schedule[day][shift].length < 2) {
                                this.schedule[day][shift].push(name);
                                data.assignedDays++;
                                break;
                            }
                        }
                    }
                }
                
            }
        }
    }
   
    resolveConflicts() {
        for (const day of this.days) {
            for (const shift of this.shifts) {
                while (this.schedule[day][shift].length > 2) {
                    const employee = this.schedule[day][shift].pop();
                    this.assignToAnotherDay(employee, day);
                }
            }
        }
    }
    
    assignToAnotherDay(employee, currentDay) {
        for (const day of this.days) {
            if (day !== currentDay && this.employees[employee].assignedDays < 5) {
                for (const shift of this.shifts) {
                    if (this.schedule[day][shift].length < 2) {
                        this.schedule[day][shift].push(employee);
                        this.employees[employee].assignedDays++;
                        return;
                    }
                }
            }
        }
    }
    


    displaySchedule() {
        const scheduleContainer = document.getElementById("schedule");
        scheduleContainer.innerHTML = "<h2>Final Employee Schedule:</h2>";
        const table = document.createElement("table");
        table.innerHTML = `<tr><th>Day</th><th>Morning</th><th>Afternoon</th><th>Evening</th></tr>`;
        
        for (const [day, shifts] of Object.entries(this.schedule)) {
            const row = document.createElement("tr");
            row.innerHTML = `<td>${day}</td><td>${shifts.morning.join(", ") || "-"}</td><td>${shifts.afternoon.join(", ") || "-"}</td><td>${shifts.evening.join(", ") || "-"}</td>`;
            table.appendChild(row);
        }
        scheduleContainer.appendChild(table);
    }
}



const scheduler = new EmployeeScheduler();

document.getElementById("addEmployee").addEventListener("click", () => {
    const name = document.getElementById("employeeName").value;
    const preferences = Array.from(document.querySelectorAll("input[name='shift']:checked")).map(input => input.value);
    if (name && preferences.length > 0) {
        scheduler.addEmployee(name, preferences);
    }
});

document.getElementById("clearButton").addEventListener("click", function() {
    document.getElementById("employeeName").value = ""; // Clears the textbox
    document.getElementById("myCheckbox").checked = false;
    document.getElementById("myCheckbox1").checked = false;
    document.getElementById("myCheckbox2").checked = false;
});

File: Assignment4.py
class EmployeeScheduler:
    def __init__(self):
        self.employees = {}
        self.schedule = {}
        self.days = ["Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday", "Sunday"]
        self.shifts = ["morning", "afternoon", "evening"]
        self.max_employees = 6
        self.initialize_schedule()

    def add_employee(self, name, preferences):
        if len(self.employees) < self.max_employees and name not in self.employees:
            self.employees[name] = {"preferences": preferences, "assigned_days": 0}
            self.assign_shifts()
            self.resolve_conflicts()
            self.display_schedule()
        else:
           
            print("No employee works more than one shift per day!(Also every employee works only 5 days per week)")
            

    def initialize_schedule(self):
        for day in self.days:
            self.schedule[day] = {shift: [] for shift in self.shifts}

    def assign_shifts(self):
        for day in self.days:
            for name, data in self.employees.items():
                if data["assigned_days"] < 5:
                    assigned = False
                    for shift in data["preferences"]:
                        if len(self.schedule[day][shift]) < 2:
                            self.schedule[day][shift].append(name)
                            data["assigned_days"] += 1
                            assigned = True
                            break
                    
                    if not assigned:
                        for shift in self.shifts:
                            if len(self.schedule[day][shift]) < 2:
                                self.schedule[day][shift].append(name)
                                data["assigned_days"] += 1
                                break

    def resolve_conflicts(self):
        for day in self.days:
            for shift in self.shifts:
                while len(self.schedule[day][shift]) > 2:
                    employee = self.schedule[day][shift].pop()
                    self.assign_to_another_day(employee, day)

    def assign_to_another_day(self, employee, current_day):
        for day in self.days:
            if day != current_day and self.employees[employee]["assigned_days"] < 5:
                for shift in self.shifts:
                    if len(self.schedule[day][shift]) < 2:
                        self.schedule[day][shift].append(employee)
                        self.employees[employee]["assigned_days"] += 1
                        return

    def display_schedule(self):
        print("\nFinal Employee Schedule:")
        print(f"{'Day':<10}{'Morning':<20}{'Afternoon':<20}{'Evening':<20}")
        print("-" * 70)
        for day, shifts in self.schedule.items():
            print(f"{day:<10}{', '.join(shifts['morning']) or '-':<20}{', '.join(shifts['afternoon']) or '-':<20}{', '.join(shifts['evening']) or '-':<20}")

if __name__ == "__main__":
    scheduler = EmployeeScheduler()
    while len(scheduler.employees) < scheduler.max_employees:
        name = input("Enter employee name (or 'stop' to finish): ")
        if name.lower() == 'stop':
            break
        preferences = input("Enter preferred shifts (morning, afternoon, evening) separated by commas: ").split(',')
        preferences = [p.strip() for p in preferences]
        scheduler.add_employee(name, preferences)
